ethtool: Add support of channel

SteamedFish opened the pull request:
Equivalent to `ethtool -l <nic_name>` command.

Example code included.

SteamedFish committed:
ethtool: Add support of channel

Equivalent to `ethtool -l <nic_name>` command.

Example code included.

DmytroLinkin commented on examples/dump_channel.rs:
```diff
@@ -0,0 +1,29 @@
+
+use futures::stream::TryStreamExt;
+
+// Once we find a way to load netsimdev kernel module in CI, we can convert this
```
s/netsimdev/netdevsim/

SteamedFish replied:
This comment is copied from other examples provided in the project.
I'll submit a dedicated PR to fix this in all the code examples.

DmytroLinkin commented on src/message.rs:
```diff
@@ -23,6 +24,8 @@ const ETHTOOL_MSG_LINKMODES_GET: u8 = 4;
 const ETHTOOL_MSG_LINKMODES_GET_REPLY: u8 = 4;
 const ETHTOOL_MSG_RINGS_GET: u8 = 15;
 const ETHTOOL_MSG_RINGS_GET_REPLY: u8 = 16;
+const ETHTOOL_MSG_CHANNELS_GET: u8 = 17;
+const ETHTOOL_MSG_CHANNELS_GET_REPLY: u8 = 18;
 const ETHTOOL_MSG_COALESCE_GET: u8 = 19;
 const ETHTOOL_MSG_COALESCE_GET_REPLY: u8 = 20;
 const ETHTOOL_MSG_TSINFO_GET: u8 = 25;
```
This const list is quite confusing, because it actually mixes values from two separate lists. Isn't it better to have "userspace to kernel" and "kernel to userspace" separately?

SteamedFish replied:
This const list is in one enum in `ethtool_netlink.h`, not from two seperate lists.

DmytroLinkin reviewed (commented):
Left few small comments.
I've tested the code - it works.

SteamedFish committed:
Merge branch 'main' into main

SteamedFish committed:
remove confusing const list

the const list comes from two different enums. Put them into two different enums
to make it clear.

cathay4t commented on src/message.rs:
```diff
@@ -15,52 +17,44 @@ use crate::{
-const ETHTOOL_MSG_TSINFO_GET: u8 = 25;
-const ETHTOOL_MSG_TSINFO_GET_REPLY: u8 = 26;
+#[repr(u8)]
+#[derive(Debug, PartialEq, Eq, Clone, Copy, IntoPrimitive, FromPrimitive)]
```
In stead of introducing new dependency, please use `From<u8>` and remove `#[repr(u8)]`.

cathay4t commented on src/message.rs:
```diff
@@ -15,52 +17,44 @@ use crate::{
+    PauseSet = 22,
+    TsInfoGet = 25,
+    #[num_enum(catch_all)]
+    UnSupport(u8),
```
Please use `Other(u8)` like other code. Yes this forbid you from using `repr[u8]`, please `impl From` trait for it when needed.

cathay4t commented on src/message.rs:
```diff
@@ -15,52 +17,44 @@ use crate::{
+    PauseGetReply = 22,
+    TsInfoGetReply = 26,
+    #[num_enum(catch_all)]
+    UnSupport(u8),
```
`Other(u8)`

cathay4t reviewed (changes requested):
Introducing new dependency should be carefully reviewed. In this case, it is not a stuff we cannot do elegantly by ourselves, hence please remove it the use of `num_enum`.

ruocco commented:
Hello, I'm interested in this feature being merged (as well as the `set` support in #14 ). 

@SteamedFish if you don't have time or have shifted priorities, do you mind if I jump in and complete this?

SteamedFish commented:
> Hello, I'm interested in this feature being merged (as well as the `set` support in #14 ).
> 
> @SteamedFish if you don't have time or have shifted priorities, do you mind if I jump in and complete this?

Sure, I got several issues managing ethernet devices in virtual machines with netlink, I've switched to ioctl.

cathay4t commented:
@SteamedFish Is that OK for me to close this PR since you moved to ioctl() approach?
